nni/nas/oneshot/pytorch/strategy.py

```diff
@@ -83,6 +85,80 @@ def export_top_models(self, top_k: int = 1) -> list[Any]:
             warnings.warn('One-shot strategy currently only supports exporting top-1 model.', RuntimeWarning)
         return [self.model.export()]
 
+    @staticmethod
+    def sub_state_dict(model, super_state_dict):
+        """
+        Truncate the state dict of the fixed subnet from that of the search space.
+
+        For example, when you already have a state dict for the base model / search space (which often
+        happens when you have trained a supernet with one-shot strategies), the state dict isn't organized
+        in the same way as when a sub-model is sampled from the search space. This patch will help
+        the modules in the sub-model find the corresponding name and param in the base model.
+        """
+        assert isinstance(model, nn.Module), 'PyTorch is the only supported framework for now.'
+        # first get the full mapping
+        full_mapping = {}
+        state_dict = model.state_dict()
+        def update_state_dict(src_prefix, tar_prefix, module):
+            if hasattr(module, STATE_DICT_PY_MAPPING):
+                # only values are complete
+                local_map = getattr(module, STATE_DICT_PY_MAPPING)
+            elif hasattr(module, STATE_DICT_PY_MAPPING_PARTIAL):
+                # keys and values are both incomplete
+                local_map = getattr(module, STATE_DICT_PY_MAPPING_PARTIAL)
+                local_map = {k: tar_prefix + v for k, v in local_map.items()}
+            else:
+                # no mapping
+                local_map = {}
+
+            if '__self__' in local_map:
+                # special case, overwrite prefix
+                tar_prefix = local_map['__self__'] + '.'
+
+            for key, value in local_map.items():
+                if key != '' and key not in module._modules:  # not a sub-module, probably a parameter
+                    full_mapping[src_prefix + key] = value
+
+            # To deal with leaf nodes.
+            for name, sub_param in itertools.chain(module._parameters.items(), module._buffers.items()):  # direct children
+                if sub_param is None or name in module._non_persistent_buffers_set:
+                    # it won't appear in state dict
+                    continue
+                if (src_prefix + name) not in full_mapping:
+                    full_mapping[src_prefix + name] = tar_prefix + name
+
+                sup_param = super_state_dict[tar_prefix + name]
+
+                assert sub_param.ndim == sup_param.ndim
+                # truncation operation of custom defined mixed module
+                if hasattr(module, "truncation"):
+                    indices = getattr(module, "truncation")(name, sub_param.shape, sup_param.shape)
+                # conv2d weight truncation
+                # TODO: groups>1 unsupported, fixe me later.
+                elif isinstance(module, nn.Conv2d) and name == "weight":
```

Comment:
I have concerns in putting Conv-specific logics here. I think it should belong to the implementation of MixedConv2d, so that it can be possibly reused in different strategies in future.

For example, we can have a `sliced_state_dict(*args)` in MixedConv2d.

Reply:
May be we should add `sliced_state_dict` method as a default method in all `MixedOperation`

Reply:
Sounds like an interface to me. Different mixed operation have different logics when processing their specific arguments. For example, Conv2d and Linear have different behaviors when processing `in_features` / `in_channels`.